File: Examples/test-suite/octave/director_langobj_runme.m
director_langobj

function self=OctCallback()
  global director_langobj;
  self=subclass(director_langobj.Callback());
  self.callback=@OctCallback_callback;
  self.callback_ref=@OctCallback_callback_ref;
end
function OctCallback_callback(self,param1)
  global received;
  received=param1;
end
function OctCallback_callback_ref(self,param1)
  global received;
  received=param1;
end

global received;

handler = OctCallback();
obj = "test data";

% An opaque proxy would arrive as a swig object rather than the string that was
% passed in.

received = 0;
director_langobj.call_callback(handler,obj);
if (!ischar(received) || !strcmp(received,obj))
  error("callback did not receive the value passed from C++")
endif

received = 0;
director_langobj.call_callback_ref(handler,obj);
if (!ischar(received) || !strcmp(received,obj))
  error("callback_ref did not receive the value passed from C++")
endif

File: Examples/test-suite/perl5/director_langobj_runme.pl
use strict;
use warnings;
use Test::More tests => 7;
use Scalar::Util qw(refaddr);
use B ();
BEGIN { use_ok 'director_langobj' }
require_ok 'director_langobj';

{
  package MyCallback;
  use base 'director_langobj::Callback';
  our $received;
  sub callback { my($self, $param1) = @_; $received = $param1; }
  sub callback_ref { my($self, $param1) = @_; $received = $param1; }
}

my $handler = MyCallback->new();
my $obj = [1, 2, 3];

# An opaque proxy would arrive as a blessed SWIG object rather than the array
# reference that was passed in.
$MyCallback::received = undef;
director_langobj::call_callback($handler, $obj);
is(ref($MyCallback::received), 'ARRAY', 'callback received an array reference');
is(refaddr($MyCallback::received), refaddr($obj), 'callback received the object itself');

$MyCallback::received = undef;
director_langobj::call_callback_ref($handler, $obj);
is(ref($MyCallback::received), 'ARRAY', 'callback_ref received an array reference');
is(refaddr($MyCallback::received), refaddr($obj), 'callback_ref received the object itself');

# The reference count of the referenced array must not drift as upcalls are made.
$MyCallback::received = undef;
my $before = B::svref_2object($obj)->REFCNT;
for (1 .. 100) {
  director_langobj::call_callback($handler, $obj);
  director_langobj::call_callback_ref($handler, $obj);
}
$MyCallback::received = undef;
my $after = B::svref_2object($obj)->REFCNT;
is($after, $before, 'reference count stable across director upcalls');
